### ISSUE TITLE
Build and run/debug VeraCrypt on Visual Studio 2010

### ISSUE THREAD

**Issue by AnhPC03** [veracrypt/VeraCrypt#1413]
Hi @idrassi, could I know how to build and run or debug the project on VS2010 or VS2019?
Because I want to change some code to print something when running the application that I can see in the console.

**Comment by idrassi** [comment]
For now, debugging can be done by opening the `VeraCrypt.sln` solution in VS2010 or VS2019 and using the "All Debug" configuration to build the projects.

Regarding printing to the console, this is not possible by default because the VeraCrypt executable is a WIN32 GUI application that is not attached to a console. Therefore, `printf` and similar functions won't work. 

However, you can programmatically allocate a console window for your WIN32 GUI application and redirect the standard I/O streams (`stdout`, `stdin`, and `stderr`) to the newly created console. This will allow you to use `printf` or `std::cout` to print logs.

Here is a function that creates a console and attaches it to the process:

```c
#include <stdio.h>
#include <fcntl.h>
#include <io.h>

void CreateConsole()
{
    // Allocate a console for this app
    if (!AllocConsole()) {
        MessageBox(NULL, "Failed to allocate console!", "Error", MB_OK | MB_ICONERROR);
        return;
    }

    FILE* fp;

    // Redirect stdout to console
    if (_wfreopen_s(&fp, L"CONOUT$", L"w", stdout) != 0) {
        MessageBox(NULL, "Failed to redirect stdout!", "Error", MB_OK | MB_ICONERROR);
        return;
    }

    // Redirect stdin to console
    if (_wfreopen_s(&fp, L"CONIN$", L"r", stdin) != 0) {
        MessageBox(NULL, "Failed to redirect stdin!", "Error", MB_OK | MB_ICONERROR);
        return;
    }

    // Redirect stderr to console
    if (_wfreopen_s(&fp, L"CONOUT$", L"w", stderr) != 0) {
        MessageBox(NULL, "Failed to redirect stderr!", "Error", MB_OK | MB_ICONERROR);
        return;
    }

    // Set console mode to support Unicode output
    SetConsoleOutputCP(CP_UTF8);
    SetConsoleCP(CP_UTF8);

    // Optionally, change the console mode to make it more usable for input
    HANDLE hConsole = GetStdHandle(STD_INPUT_HANDLE);
    DWORD mode;
    if (GetConsoleMode(hConsole, &mode)) {
        // Disable Quick Edit mode if needed
        SetConsoleMode(hConsole, mode & ~(ENABLE_QUICK_EDIT_MODE));
    }

    // Ensure std::wcout works correctly
    std::ios::sync_with_stdio(true);
    std::wcout.clear();
    std::wcin.clear();
    std::wcerr.clear();
}
```

You can then call `CreateConsole` in the `InitApp` function defined in `Dlgcode.c`:

```c
void InitApp (HINSTANCE hInstance, wchar_t *lpszCommandLine)
{
    WNDCLASSW wc;
    char langId[6];    
    SetDefaultDllDirectoriesPtr SetDefaultDllDirectoriesFn = NULL;
#if !defined(SETUP)
    wchar_t modPath[MAX_PATH];
#endif
    INITCOMMONCONTROLSEX InitCtrls;

    InitOSVersionInfo();

    CreateConsole(); // <-- Added

    // ... rest of the code
}
```

**Comment by AnhPC03** [comment]
@idrassi Thank you very much for your very detail answer. I'll try as your guide.

**Comment by AnhPC03** [comment]
Hi @idrassi, I did it again but got the error as in the picture. I chose `All Debug` and `x64`. I ran with `Debug -> Start Without Debugging`
<img width="1347" alt="Screen Shot 2024-09-09 at 22 35 59" src="https://github.com/user-attachments/assets/4a2b323c-af9c-47a7-97f3-1db6802c703f">

**Comment by idrassi** [comment]
In your screenshot, I see that the "Setup" project is in bold, which indicates that it is the default project when debugging. However, you cannot debug the Setup project this way because the final setup, which is distributed, must first be created from the setup generated using the `/p` switch.

To debug the setup installer, you need to build all binaries in debug mode (x86, x64, and ARM64), then run the batch file "Signing\sign_test_debug.bat". This batch file will sign the binaries using a test certificate and create a setup installer that can be debugged. You can read the batch file to get better understanding about the process.

The generated debug installer can then be debugged (Note: Visual Studio must be run as Administrator to debug the setup).

**Comment by idrassi** [comment]
No, you don't need to boot Windows into test mode to use the test binaries.

The only exception is if you rebuild the driver project, which is disabled by default. Officially signed drivers are included in the source code repository (under `src\Release\Setup Files`).

So, simply rebuild the solution in debug mode, run `sign_test_debug.bat`, and the generated setup installer can be used on a normal Windows system.

**Comment by AnhPC03** [comment]
@idrassi Thank you very much. I'll try it again

**Comment by AnhPC03** [comment]
@idrassi Hello, could you help me with this?
I saw the default Platform toolset is "Windows7.1SDK" but building solution failed because there is no "Windows7.1SDK" value in the Platform toolset list
![Screen Shot 2024-09-12 at 10 51 44](https://github.com/user-attachments/assets/373bf9bb-c958-4c6e-91d3-edb934d45845)

**Comment by idrassi** [comment]
You can change it v100. It is not mandatory to use Windows SDK 7.1.
You can refer to the Windows Build Guide in the documentation: https://veracrypt.fr/en/CompilingGuidelineWin.html

**Comment by AnhPC03** [comment]
@idrassi I installed and ran successfully all steps in Windows Build Guide including executing "sign_test.bat" and importing certificates but after that I sill couldn't install VeraCrypt by "VeraCrypt Setup 1.26.15.exe" file.  The error when I clicked the setup file
![Screen Shot 2024-09-12 at 14 18 48](https://github.com/user-attachments/assets/6ab6cb78-245a-4a52-988f-470fc90047e4)

There was one step that I hadn't tried yet
`Compute SHA512 fingerprint of the test code signing certificate and update the gpbSha512CodeSignCertFingerprint array in the file "src/Common/Dlgcode.c" accordingly.`

Could you explain how can I compute sha512 fingerprint off the test code signing certificate?

**Comment by sergeevabc** [comment]
@AnhPC03, are you building VeraCrypt 1.26.15 for Windows from sources? Oh, I just took a pre-built binary from Sourceforge repository and works fine on Windows 7. What actually troubles me is that M. Idrassi all of a sudden cut users like me off.

![Image](https://github.com/user-attachments/assets/7b8dd71d-6381-4044-9f64-d564b27b47e9)

WTF is this? Betrayal.
